### PR TITLE
country, city, district, hotel, room

### PR DESCRIPTION
**Pull request by wmosienski**

### CODE REVIEW

File: src/service/log/log.service.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import Log from '../../datastructure/log/log.interface';
+import FileRepository from '../../repository/fileRepository/file.repository';
+import MongoLogRepository from '../../repository/mongoRepository/entities/mongo.log.repository';
+import moment from 'moment';
+
+const logsPath = './logs/';
+const timeFormat = 'YYYY_MM_DD';
+const getFilePath = () => logsPath + moment().format(timeFormat) + '.txt';
+const charSize = 8; // UTF-8
+const separator = '\n';
```
Comment (Kolya-kostevich) [inline src/service/log/log.service.ts:10]: You should use here constant from path library, like  separator

File: src/service/log/log.service.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import Log from '../../datastructure/log/log.interface';
+import FileRepository from '../../repository/fileRepository/file.repository';
+import MongoLogRepository from '../../repository/mongoRepository/entities/mongo.log.repository';
+import moment from 'moment';
+
+const logsPath = './logs/';
+const timeFormat = 'YYYY_MM_DD';
+const getFilePath = () => logsPath + moment().format(timeFormat) + '.txt';
```
Comment (Kolya-kostevich) [inline src/service/log/log.service.ts:8]: const date =  moment().format(timeFormat);

Comment (Kolya-kostevich) [reply src/service/log/log.service.ts:8]: () => {
const date = moment().format(timeFormat);

return  logsPath + date  + '.txt';
}

File: src/service/log/log.service.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    static log(message: string, level?: number, meta?: object) {
+        return LogService.customLog({
+            date: moment().format('YYYY:MM:DD:HH:mm:ss:SS'),
+            level: level || 1,
+            message,
+            meta: meta || {},
+        })
+    }
+
+    static async customLog(log: Log | any) {
```
Comment (Kolya-kostevich) [inline src/service/log/log.service.ts:36]: addLog

Comment (Kolya-kostevich) [reply src/service/log/log.service.ts:36]: You should use 'Any' type of Typescript only in edge cases, I think here you can get rid of it.

File: src/service/log/log.service.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            await LogService.mongoLogRepository.insertBatch(LogService.logs);
+            LogService.fileRepository.clearFile();
+            LogService.lastFileSize = logSize;
+            LogService.logs = [];
+        }
+
+        const updatedFilePath = getFilePath();
+        if (LogService.lastFilePath !== updatedFilePath) {
+            LogService.fileRepository.end();
+            LogService.fileRepository = new FileRepository(updatedFilePath);
```
Comment (Kolya-kostevich) [inline src/service/log/log.service.ts:51]: You can move it to separate method like CreateNewFileRepository

Comment (Kolya-kostevich) [reply src/service/log/log.service.ts:51]: Rename FileRepository LogsStorage

File: src/service/room/room.service.ts [hunk 5]
```diff
@@ -1,21 +1,52 @@
...
+import { classLog } from '../../utils/log/log';
 
-class RoomModel {
-    room: Room;
+const repository = new MongoRoomRepository();
+const hotelRepository = new MongoHotelRepository();
 
-    constructor(room: Room) {
-        this.room = room;
+@classLog([])
```
Comment (Kolya-kostevich) [inline src/service/room/room.service.ts:11]: Use First Letter in uppercase

File: src/service/room/room.service.ts [hunk 6]
```diff
@@ -1,21 +1,52 @@
...
+class RoomService {
+
+    static async create(id: string, room: Room) {
```
Comment (Kolya-kostevich) [inline src/service/room/room.service.ts:14]: Tell please why hasn't been added id into the room

File: src/service/room/room.service.ts [hunk 7]
```diff
@@ -1,21 +1,52 @@
...
+        const errors = errorInfo({});
```
Comment (Kolya-kostevich) [inline src/service/room/room.service.ts:15]: Please rename it like init error info

File: src/service/room/room.service.ts [hunk 8]
```diff
@@ -1,21 +1,52 @@
...
+        const validationResult = roomValidate(room);
```
Comment (Kolya-kostevich) [inline src/service/room/room.service.ts:16]: Rename it to validateRoom

File: src/service/room/room.validate.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
+import Room from "../../datastructure/room/room.interface";
+import { ErrorInfo, isErrorInfo } from "../../utils/error/error";
+import { validateIsNumber, validateIsPositive } from "../../utils/error/validation/number.error";
+
+export const roomNumberValidate = (name: number): ErrorInfo[] => [
```
Comment (Kolya-kostevich) [inline src/service/room/room.validate.ts:5]: it should be looks like you call validate function and inside of function you should have (!validateIsNumber) {
     error.message = 'It's not a number'
}

Comment (Kolya-kostevich) [inline src/service/room/room.validate.ts:5]: Please rename 'ErrorInfo' to 'ValidationResult'

File: src/utils/error/validation/number.error.ts [hunk 10]
```diff
@@ -0,0 +1,14 @@
+import { isNumber, isPositive } from "../../validation/number.validation";
+import { ErrorInfo, errorInfo } from "../error";
+
+export const validateIsPositive = (value: number): ErrorInfo | any => {
```
Comment (Kolya-kostevich) [inline src/utils/error/validation/number.error.ts:4]: Rename it to validateIsPositiveNumber

File: src/repository/mongoRepository/entities/mongo.city.repository.ts [hunk 11]
```diff
@@ -0,0 +1,53 @@
+import mongoose from "mongoose";
+import CityMongoModel from "../../../database/mongo/schema/city.schema";
+import City from "../../../datastructure/city/city.interface";
+import DataIdentifierBase from "../../../datastructure/data.identifier.base";
+import DataResultBase from "../../../datastructure/data.result.base";
+import { classLog } from "../../../utils/log/log";
+import MongoRepository from "../mongo.repository";
+
+export default class MongoCityRepository extends MongoRepository {
```
Comment (Kolya-kostevich) [inline src/repository/mongoRepository/entities/mongo.city.repository.ts:9]: It should be like *CityRepository*

File: src/service/log/log.service.ts (hunk 1 above)
Comment (wmosienski) [reply src/service/log/log.service.ts:10]: little misunderstanding, that was meant to be line separator which is ok to be '\n'. however Im gonna change the name to 'lineSeparator' and move it to the actual file handler (fileRepostiroty - but its name is going to be chaged as well due to your comments below)

File: src/service/log/log.service.ts (hunk 2 above)
Comment (wmosienski) [reply src/service/log/log.service.ts:8]: addressed

File: src/service/log/log.service.ts [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+const separator = '\n';
+
+class LogService {
+    static logs: Log[] = [];
+    static lastFilePath: string = getFilePath();
+    static mongoLogRepository = new MongoLogRepository();
+    static fileRepository = new FileRepository(getFilePath());
+    static lastFileSize: number;
+    static options = {
+        maxFileSize: 1950, //1024*50,
```
[the remark this answers is not in the document]
Comment (wmosienski) [reply src/service/log/log.service.ts:19]: 1950 was for testing purposes. Im going to make const for 1024*2 (2kb)

File: src/service/log/log.service.ts [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+    static init() {
+        LogService.fileRepository.start();
+        LogService.lastFileSize = charSize * (LogService.fileRepository.getCharLength() || 0);
+    }
+
+    static log(message: string, level?: number, meta?: object) {
+        return LogService.customLog({
+            date: moment().format('YYYY:MM:DD:HH:mm:ss:SS'),
+            level: level || 1,
```
[the remark this answers is not in the document]
Comment (wmosienski) [reply src/service/log/log.service.ts:30]: addressed

File: src/service/log/log.service.ts (hunk 3 above)
Comment (wmosienski) [reply src/service/log/log.service.ts:36]: addressed

File: src/service/log/log.service.ts [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+        const updatedFilePath = getFilePath();
+        if (LogService.lastFilePath !== updatedFilePath) {
+            LogService.fileRepository.end();
+            LogService.fileRepository = new FileRepository(updatedFilePath);
+            LogService.init();
+        }
+        LogService.fileRepository.append(logString + separator);
+    }
+
+    static logToString(log: Log) {
```
[the remark this answers is not in the document]
Comment (wmosienski) [reply src/service/log/log.service.ts:57]: addressed

File: src/service/log/log.service.ts (hunk 4 above)
Comment (wmosienski) [reply src/service/log/log.service.ts:51]: addressed (moved to provider folder and renamed)

File: src/service/room/room.service.ts (hunk 5 above)
Comment (wmosienski) [reply src/service/room/room.service.ts:11]: addressed

File: src/service/room/room.service.ts (hunk 8 above)
Comment (wmosienski) [reply src/service/room/room.service.ts:16]: ok, now I remember, it's a hotel ID which is required to create room (a room exists in a particular hotel). it's a naming issue, I'll rename those ids. in this case it'll be hotelID

File: src/utils/error/validation/number.error.ts (hunk 10 above)
Comment (wmosienski) [reply src/utils/error/validation/number.error.ts:4]: addressed

File: src/service/room/room.validate.ts (hunk 9 above)
Comment (wmosienski) [reply src/service/room/room.validate.ts:5]: addressed

File: src/service/room/room.service.ts (hunk 7 above)
Comment (wmosienski) [reply src/service/room/room.service.ts:15]: addressed

File: src/service/room/room.service.ts (hunk 6 above)
Comment (wmosienski) [reply src/service/room/room.service.ts:14]: ok, now I remember, it's a hotel ID which is required to create room (a room exists in a particular hotel). it's a naming issue, I'll rename those ids. in this case it'll be hotelID

File: src/service/room/room.validate.ts (hunk 9 above)
Comment (wmosienski) [reply src/service/room/room.validate.ts:5]: I disagree, the exact logic you suggested is inside 'validateIsNumber' and may be reused by other validators like here it is used. moreover writing only [validateIsNumber, validateIsPositiveNumber] is more readible than any if statements. pls correct my reasoning.

File: src/controller/controllers/city/city.controller.ts [hunk 15]
```diff
@@ -23,7 +23,7 @@ class CityController extends ControllerBase {
     }
 
     async getAll(req: Request, res: Response, next: NextFunction) {
-        const errors = errorInfo({});
+        const errors = createValidationResult({});
```
Comment (Kolya-kostevich) [inline src/controller/controllers/city/city.controller.ts:26]: const emtpty_validation_result

**Pull request merged**